lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -271,50 +291,88 @@ applyBlock
+-- > bs = [b1, b2, …, bn]@
 --
--- > b = [b1, b2, ..., bn]
+-- , the function returns
 --
--- Returns the following list of updates:
---
--- > [ (t b1, w + b1)
--- > , (t b2, w + b1 + b2)
--- > , ...
--- > , (t bn, w + b1 + b2 + ... + bn) ]
+-- @
+-- [ (filtered b1, (delta w0 -> w1 , w1 = w0 + b1))
+-- , (filtered b2, (delta w1 -> w2 , w2 = w1 + b2))
+-- , …
+-- , (filtered bn, (delta w(n-1) -> wn, wn = w(n-1)+bn))
+-- ]
 --
--- Where:
+-- Here:
 --
--- * __@(t bi)@__   is the set of transactions contained within block __@bi@__.
--- * __@(w + bi)@__ is the wallet state after applying block __@bi@__ to wallet
---   __@w@__.
+-- * @filtered bj@ refers to the set of transactions contained in the
+--   block @bj@ that were actually applied to the wallet state.
+-- * @wi + bj@ refers to the wallet state obtained after applying
+--   the block @bi@ to the wallet @wj@.
+-- * delta wi -> wj@ refers to the delta that was applied in order
+--   to obtain @wj@ from @wi@.
 --
 applyBlocks
     :: (IsOurs s Address, IsOurs s RewardAccount, Monad m)
     => BlockData m (Either Address RewardAccount) ChainEvents s
     -> Wallet s
-    -> m (NonEmpty (FilteredBlock, (DeltaWallet s, Wallet s)))
-applyBlocks (List (block0 :| blocks)) cp = pure $
-    NE.scanl (flip applyBlock . snd . snd) (applyBlock block0 cp) blocks
-applyBlocks (Summary _ _) _ = error "FIXME: Implement me!"
-    -- FIXME: Return type (NonEmpty of filtered blocks)
-    -- needs to be rethought, as it still assumes sequential order
-    -- of blocks.
+    -> m (NonEmpty ([FilteredBlock], (DeltaWallet s, Wallet s)))
+applyBlocks (List (block0 :| blocks)) w0 = pure $
+    NE.scanl applyBlock' (first (:[]) $ applyBlock block0 w0) blocks
+  where
+    applyBlock' (_,(_,w)) block = first (:[]) $ applyBlock block w
+applyBlocks summary@(Summary _ _) w =
+    (NE.:| []) <$> applyBlockData summary w
+
+-- | Apply multiple blocks in sequence to an existing wallet
+-- and return the final wallet state as well as the transactions
+-- that were applied.
+applyBlockData
+    :: (IsOurs s Address, IsOurs s RewardAccount, Monad m)
+    => BlockData m (Either Address RewardAccount) ChainEvents s
+    -> Wallet s
+    -> m ([FilteredBlock], (DeltaWallet s, Wallet s))
+applyBlockData blocks (Wallet !u0 _ s0) = do
+    (chainEvents, s1) <- discoverFromBlockData blocks s0
+    let blockEvents = toAscBlockEvents chainEvents
+        applies u blockEvent = applyBlockEventsToUTxO blockEvent s1 u
+        (processedBlocks, u1) = mapAccumL' applies u0 blockEvents
+        filteredBlocks = map fst processedBlocks
+        tip1 = lastHeader blocks
+        dtip = Delta.Replace tip1
+        ds = Delta.Replace s1
+        du = mconcat (reverse $ map snd processedBlocks)
+        dw = DeltaWallet
+            { deltaUTxO = du , deltaAddressBook = ds, deltaCurrentTip = dtip }
+    pure (filteredBlocks, (dw, Wallet u1 tip1 s1))
+
+-- | Strict variant of 'mapAccumL'.
+mapAccumL' :: (s -> a -> (o,s)) -> s -> [a] -> ([o],s) 
```

Comment:
when I see something like this I always would like to make sure that even irrespective of bang patterns applied we have no thunk build up. I wonder if we should not start using `nothunk` - some reading for inspiration https://well-typed.com/blog/2020/09/nothunks/

Reply:
Probably. In other parts of our code, `nothunks` is used, but not here yet. I do want to investigate the space usage of `Wallet` and `WalletState` more systematically at some point, but not right now — things are working out ok at the moment, so I figured that I could away with just making sure that I don't accidentally make things worse for the time being.